Created batch-file to create a windows executable with pyinstaller

Hakuna75 opened the pull request:
I've created a batch-file that can create a Windows executable for git-restore-mtime using pyinstaller.

This way you only need this executable; you don't need to install python or any libraries.

I'm using this myself in my Git repositories; it might just be useful to other people too.

If you want me to change anything to conform to any standards let me know.

MestreLion commented on windows/build_windows_executable.bat:
```diff
@@ -0,0 +1,22 @@
+ECHO Testing script...
+python ..\git-restore-mtime --help
+
+REM Bij gebruik van upx is er een vc140dll runtime error
```
What?

Hakuna75 commented:
Sorry, dutch :D
'Bij gebruik van upx is er een vc140dll runtime error' ~= 'When using upx a vc140dll runtime error occurs when running the created executable'

I'll fix the comment.

Regards,

Remco

________________________________
From: Rodrigo Silva <notifications@github.com>
Sent: Thursday, November 21, 2019 08:47
To: MestreLion/git-tools <git-tools@noreply.github.com>
Cc: Hakuna75 <github@voogt.nu>; Author <author@noreply.github.com>
Subject: Re: [MestreLion/git-tools] Created batch-file to create a windows executable with pyinstaller (#35)


@MestreLion commented on this pull request.

________________________________

In windows/build_windows_executable.bat<https://github.com/MestreLion/git-tools/pull/35#discussion_r348933964>:

> @@ -0,0 +1,22 @@
+@ECHO OFF
+REM pip install --upgrade --trusted-host pypi.org --trusted-host files.pythonhosted.org pip setuptools
+REM pip install --trusted-host pypi.org --trusted-host files.pythonhosted.org https://github.com/pyinstaller/pyinstaller/archive/develop.tar.gz
+PUSHD "%~dp0"
+CD
+
+ECHO.
+ECHO Checking current python.exe location(s)
+where python.exe
+
+ECHO.
+ECHO Testing script...
+python ..\git-restore-mtime --help
+
+REM Bij gebruik van upx is er een vc140dll runtime error


What?

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/MestreLion/git-tools/pull/35?email_source=notifications&email_token=AAL2ATNCH5Q53BU32QXDW2LQUY4KFA5CNFSM4JPB3DJ2YY3PNVWWK3TUL52HS4DFWFIHK3DMKJSXC5LFON2FEZLWNFSXPKTDN5WW2ZLOORPWSZGOCMOVHWY#pullrequestreview-320689115>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAL2ATIT6267NUB6OZ2CIQ3QUY4KFANCNFSM4JPB3DJQ>.

MestreLion commented:
Thanks for sharing those!

But I wonder what's the use case of a packager to `.exe`  if the packager itself requires Python and Pip installed. It's no benefit to the end-user, since to create his own executable he would need to install python anyway.

It could be useful if _we_ generate the `git-restore-mtime.exe` binary and then host it, here or somewhere else, ready to download and use. Currently I don't have a Windows environment to do so, but It's a good idea for the future. In this case I believe it's a mini "guide" is better than the scripts, to document the instructions in a `README.Windows.md`, so I can post them when I publish the files.

Hakuna75 commented:
Hi Rodrigo,

For a single user on a single machine the benefit might be limited indeed.

I'm working in a team with multiple developers; not everyone has Python installed. This way the can use the functionality without the need to install Python.

If you want, I can also add the currently created git-restore-mtime.exe to the repository somewhere; if you have another location to host it that's ok too.

A README.Windows.md is a good idea too I guess. For myself I created the batch-files, because I like one-click solutions for initialization and building.

Regards,

Remco

________________________________
From: Rodrigo Silva <notifications@github.com>
Sent: Thursday, November 21, 2019 09:24
To: MestreLion/git-tools <git-tools@noreply.github.com>
Cc: Hakuna75 <github@voogt.nu>; Author <author@noreply.github.com>
Subject: Re: [MestreLion/git-tools] Created batch-file to create a windows executable with pyinstaller (#35)


Thanks for sharing those!

But I wonder what's the use case of a packager to .exe if the packager itself requires Python and Pip installed. It's no benefit to the end-user, since to create his own executable he would need to install python anyway.

It could be useful if we generate the git-restore-mtime.exe binary and then host it, here or somewhere else, ready to download and use. Currently I don't have a Windows environment to do so, but It's a good idea for the future. In this case I believe it's a mini "guide" is better than the scripts, to document the instructions in a README.Windows.md, so I can post them when I publish the files.

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/MestreLion/git-tools/pull/35?email_source=notifications&email_token=AAL2ATNZJKACJBEV7OJ5Y4TQUZASHA5CNFSM4JPB3DJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEEZL3HQ#issuecomment-556973470>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAL2ATN6RC5VVQFWMAYG4P3QUZASHANCNFSM4JPB3DJQ>.

MestreLion commented:
So you're the team builder and distributor... ok, that's a good use-case. So how about this structure?
- `windows/` subfolder, like you did
- `windows/.gitignore`, with no changes I guess
- `windows/README.md`, explaining about Pyinstaller and the purpose of this folder. I's important to note this is not _required_ to run `git-restore-mtime` on Windows, it's just a convenience to generate a self-contained, native binary executable if one wants to distribute it to users without Python installed. Also add a few links on where to download and install the requirements, such as Python and Pip for windows.
- `windows/build.bat`, as a single script. It should check if `python` and `pip` are installed (and fail with a message if not), upgrade pip and setuptools (better to use `--user` so no admin privileges are required), pip-install `pyinstaller` (also as user), and perform the build, exiting with a message indicating where is the generated `.exe` if the build went ok.

If you're up for this, I truly appreciate and will gladly merge!

Hakuna75 commented:
I see you already edited and integrated the other changes, thanks!

I've rebased the windows_executable branch on your master branch and I'll continue with your suggested changes.

Regards,

Remco

________________________________
From: Rodrigo Silva <notifications@github.com>
Sent: Thursday, November 21, 2019 21:12
To: MestreLion/git-tools <git-tools@noreply.github.com>
Cc: Hakuna75 <github@voogt.nu>; Author <author@noreply.github.com>
Subject: Re: [MestreLion/git-tools] Created batch-file to create a windows executable with pyinstaller (#35)


So you're the team builder and distributor... ok, that's a good use-case. So how about this structure?

  *   windows/ subfolder, like you did
  *   windows/.gitignore, with no changes I guess
  *   windows/README.md, explaining about Pyinstaller and the purpose of this folder. I's important to note this is not required to run git-restore-mtime on Windows, it's just a convenience to generate a self-contained, native binary executable if one wants to distribute it to users without Python installed. Also add a few links on where to download and install the requirements, such as Python and Pip for windows.
  *   windows/build.bat, as a single script. It should check if python and pip are installed (and fail with a message if not), upgrade pip and setuptools (better to use --user so no admin privileges are required), pip-install pyinstaller (also as user), and perform the build, exiting with a message indicating where is the generated .exe if the build went ok.

If you're up for this, I truly appreciate and will gladly merge!

—
You are receiving this because you authored the thread.
Reply to this email directly, view it on GitHub<https://github.com/MestreLion/git-tools/pull/35?email_source=notifications&email_token=AAL2ATODXUBXDOSEMTJHSPTQU3TTDA5CNFSM4JPB3DJ2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEE3QCXA#issuecomment-557252956>, or unsubscribe<https://github.com/notifications/unsubscribe-auth/AAL2ATLRHJ2VOWYJEBSHLHLQU3TTDANCNFSM4JPB3DJQ>.

Hakuna75 commented:
I've added a README.md and updated the batch file; I've included a possibility for an automatic install of all requirements by using a /INIT-parameter for the batch file. Let me know if you want anything added/changed before you want to perform the pull request.

MestreLion commented:
Sweet!!! I loved the `README.md`, it's very comprehensive, thanks a lot!

The batch file is also great, the `/INIT` option was a sweet idea.

A small remarks/suggestion: Is there a way for windows batch files to check exit code of commands? This way your tests for pip, python, etc, can actually test if they succeeded, and abort the script with a message otherwise. On Linux (bash) this is something like this:
```
if ! python --version &> /dev/null; then
    echo "Python is not installed. Run '$0 /INIT' as administrator to install dependencies"
    exit 1
fi
```
Note that I suppressed `python --version` output, as I'm already printing a message in case of failure, so a silent run means success.

If you can/want to do this change I can wait, if not I can merge this PR right now.

And, once again, thanks a lot for your work and contribution!

Hakuna75 commented:
With Windows batch files it is possible to check exit codes of commands, but...
it's not as straightforward/elegant/consistent as a Linux bash shellscript.

It's usually done with 'IF ERRORLEVEL 1 (..)' or something like 'IF %ERRORLEVEL% eq 1'. Somehow it always takes me a couple of tries before it seems to do what I want it to; sometimes later I still find out there is a scenario that doesn't work properly. Sometimes it works better to check the actual output of a command than to interpret an exit code.

Your bash example translated to Windows batch would probably be something like this:
```
python.exe --version > NUL 2>&1
IF ERRORLEVEL 1 THEN (
  ECHO Python is not installed. Run "%~dp0" /INIT as administrator to install dependencies
  GOTO :EOF
)
```

I'll give it a try, but it can take a while before I've got something for a new pull request.

MestreLion commented:
Ok, thanks for the feedback. I've already merged, so feel free to improve it later if you want to. As it is it's already a great contributions, thank you!

Hakuna75 commented:
Great, thanks! I'll let you know when I have some improvements to the batch file.
